### ISSUE TITLE
Block/ Item system is awful

### ISSUE THREAD

**Issue by Hopson97** [Hopson97/MineCraft-One-Week-Challenge#29]
If I want to add a new block I must:

1. Add it to the block enum
2. Add it to the item enum
3. Add it to the item-block enum conversion functions
4. Add it to the block database
5. Add it to the Blocks/ folder

This is pretty awful, but may be forgivable seeing as I only had a week.

But of course, there probably is a better way, but idk

**Comment by OperationDarkside** [comment]
In consideration of modding I propose a simple solution.

Xml, Json would be easy for non-programmers to edit, but adds more dependencies.
POD would be effective for reading programmatically, but adds complexity for modders. You'd need to program a special editor. (overkill imo)

CSV is easy to edit (Notepad/Excel/Calc) and easy to load. To decrease loading time, there should only be 2 files. Meta data and image data. You could load and process both concurrently, then create the class instances afterwards.

**Comment by OperationDarkside** [comment]
Hi,

I roughly sketched what I meant in this little gist: https://gist.github.com/OperationDarkside/103a58c9dbc19f97cad70d9a7ea1e831

You might loose SOME performance with this dynamic solution, but it is extendable.

Hope it helped.

EDIT:
Reworked the BlockDB a few times. Shouldn't throw errors anymore, but building it into the system would take a lot of time. Could someone else please do that? I lack the insight and time.

**Comment by OperationDarkside** [comment]
@Hopson97 Why did you separate Blocks and Materials? In my current rework I will merge their members, because in creative mode you'd want every block (except air).

**Comment by Hopson97** [comment]
I don't really remember.

I mean, it made sense at the time , but then again, I was rushing to complete it in a week :P

**Comment by OperationDarkside** [comment]
@Hopson97 My first result with the new block loading system.
Looks like the dirt blocks are missing, but I didn't expect it to go so fast

![mcclonenewblocksystem](https://user-images.githubusercontent.com/25718228/31867654-e209607a-b793-11e7-9700-7c734e24576f.jpg)

**Comment by IridescentRose** [comment]
Woah, so does it load in any particular order? Or just gets the job done?

**Comment by OperationDarkside** [comment]
[BlockDB.csv.txt](https://github.com/Hopson97/MineCraft-One-Week-Challenge/files/1412798/BlockDB.csv.txt)


For the majority of the program there are no greater changes. Only the block loading is more dynamic and I removed the Material class.

This is the block class:
```C++
class Block {
public:
	bool isOpaque;
	bool isCollidable;
	bool isBlock;
	unsigned id; // Identifies a block type
	int maxStackSize;
	BlockMeshType meshType;
	std::string name;
	sf::Vector2i texTopCoord;
	sf::Vector2i texSideCoord;
	sf::Vector2i texBottomCoord;
};
```

The essentiall blocks are hard coded, so I can call them by id and not by name.
```C++
Block block_nothing = {true, false, false, 0, 0, static_cast<BlockMeshType>(0), "Nothing", {0, 0}, {0, 0}, {0, 0}};
    Block block_air = {true, false, false, 1, 0, static_cast<BlockMeshType>(0), "Air", {0, 0}, {0, 0}, {0, 0}};
    Block block_water = {true, false, false, 2, 0, static_cast<BlockMeshType>(0), "Water", {8, 0}, {8, 0}, {8, 0}};
```

Every other block is located in the csv file (attachment, remove the ".txt")

And that's the storage:
```C++
// dynamic storage for moddability
std::vector<Block> blocks;
// hashmap for faster named access
std::unordered_map<std::string, Block*> name_block_rel;
// hashmap for faster id access
std::unordered_map<unsigned, Block*> id_block_rel;
```

And this is how you get a block by name:
```C++
ChunkBlock DesertBiome::getUnderWaterBlock(Rand& rand) const
{
    return BlockDB::get()["Sand"].id;
}
```

**Comment by Hopson97** [comment]
I always get worried about performance when I see an std::string used as a key for a hash map, hence why I went for enum hard code approach really  :P 


But eh, we'll see. No way to know without actually trying it, really.

**Comment by OperationDarkside** [comment]
Hi,

sorry for the delayed response. Business trip and sickness...
Found the Error. The vector did a reordering, so the pointers were invalidated.
Switched to integer indexing. Now nearly everything works and I can't "feel" a major difference, but F3 doesn't work. 

![zwischestand mcclone](https://user-images.githubusercontent.com/25718228/32150775-3f7dd334-bd17-11e7-83db-a0f51d8291a8.jpg)

Sadly I'm working on an old version and I don't have a clue how to make a pull request. After fixing F3 I will update to the latest version and issue a pull request.

Happy Halloween.

**Comment by UnderTrack** [comment]
Well, for the pull request it's simple :
Go on this git, make a fork, then upload to your fork and on your fork click to create a pull request, however, you may take care to redownload the current version, put your changes into it before uploading and making the pull request.
And Welcome back :)

**Comment by IridescentRose** [comment]
I thought F3 was okay though...

**Comment by kennyrkun** [comment]
@NT-Bourgeois-Iridescence-Technologies, he somehow broke it in her quest at fixing the item system.

**Comment by OperationDarkside** [comment]
@NT-Bourgeois-Iridescence-Technologies Yes, @kennyrkun is right. The by F3 enabled showed text relied on the Material-enum, which I replaced. So I somehow messed up creating the string, which should be shown.

**Comment by OperationDarkside** [comment]
So, made the fork, applied my changes and it works. At least for me. Before I make a pull request, could someone else look over the code? I have a wierd issue, that the bool for being on the ground switches very often. Don't know, if this is related to my changes. Frames seem fine though.

**Comment by IridescentRose** [comment]
I might check it out.

**Comment by OperationDarkside** [comment]
@NT-Bourgeois-Iridescence-Technologies Thank you. I might add a FPS-Limiter too.

**Comment by kennyrkun** [comment]
@OperationDarkside that should only be an option in the configuration, similar to Minecraft's option.

**Comment by OperationDarkside** [comment]
@kennyrkun that was the plan. A config option

**Comment by IridescentRose** [comment]
@OperationDarkside I'll probably do that in the options cuz I've changed so much about how options and stuff work XD

**Comment by OperationDarkside** [comment]
Added a FPS Limiter like this. It doesn't really create the target framerate, but 200 instead of 700 fps.

```C++
const static long m_fps = 120;
constexpr const static long m_max_frametime = ((float)1 / (float)m_fps) * 1000000;

// FPS Limiting
if(m.asMicroseconds() < m_max_frametime){
     sf::sleep(sf::microseconds(m_max_frametime - m.asMicroseconds()));
}
```

**Comment by IridescentRose** [comment]
Um... @OperationDarkside you could just use the renderwindow method of sf::RenderWindow::setFrameRateLimit(int fps) to set an exact limit and disable with fps = 0, or sf::RenderWindow::setVerticalSyncEnabled(bool vsync) for an OS controlled version. I'll add this on my fork.

**Comment by OperationDarkside** [comment]
@NT-Bourgeois-Iridescence-Technologies Again something for the settings menu, because framelimiters and vsync can cause problems. But since the blocksystem remake I have that strange issue, that everytime I move the mouse to the side, the picture jumps a few degrees up. Since I don't know what could cause this, because moving up and down works fine. Maybe you as the graphics guy know something.

**Comment by IridescentRose** [comment]
@OperationDarkside I resolved that in #63 and will be uploading by this weekend. Essentially it's due to the "fix" for #61/#62 which does not properly reset the y axis. 

Also, do you mind if I pull your fork into mine? There seems to be a lot of merge conflicts to resolve, and I'd be willing to help with that. Of course I'll separate the final pull request into yours and mine in one request, because you deserve the credit for this awesome sauce 😃

**Comment by OperationDarkside** [comment]
Pull as much as you want. If the last mentioned separation is too complicated, just leave it out.
